File: day06/part2.py
import functools

_TESTCASE = """\
abc

a
b
c

ab
ac

a
a
a
a

b
""".strip().split(
    "\n\n"
)


def compute(data):
    """
    >>> compute(_TESTCASE)
    6
    """
    return sum(
        len(
            functools.reduce(
                set.__and__,
                map(set, group.splitlines()),
            )
        )
        for group in data
    )


def main():
    import pathlib

    input_path = pathlib.Path(__file__).with_name("input.txt")

    with input_path.open() as f:
        print(compute(f.read().strip().split("\n\n")))


if __name__ == "__main__":
    main()

File: day06/part2.ts
import * as path from "path";
import { readFileSync } from "fs";

function compute(groups: String[]): number {
  return groups
    .map(
      (group) =>
        group
          .split("\n")
          .map((answers) => new Set(answers))
          .reduce(
            (answersLeft, answersRight) =>
              new Set([...answersLeft].filter((a) => answersRight.has(a)))
          ).size
    )
    .reduce((a, b) => a + b);
}

function main() {
  const inputPath = path.resolve(__dirname, "input.txt");
  const input = readFileSync(inputPath, "utf-8");
  console.log(compute(input.trim().split("\n\n")));
}

main();
